File: Consumatore.java
import java.io.FileReader;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.io.FileNotFoundException;

public class Consumatore {
	public static void main(String[] args) {
		BufferedReader bufferedReader = null;
		boolean charIsPresent = false;
		String characters_to_remove;
		int nread;
		char read_char;
		
		// Controllo numero argomenti
		if (args.length < 1 || args.length > 2) {
			System.out.println("Il numero di argomenti in input è errato.\n");
			System.exit(0);
		}

		characters_to_remove = args[0];

		if (args.length == 1) { // Nel caso in cui passo soltanto i caratteri come argomento e ridirigo il file di testo in input
			bufferedReader = new BufferedReader(new InputStreamReader(System.in));
		}
		else if (args.length == 2) { // Nel caso in cui passo sia i caratteri che il nome del file negli argomenti
			try {
				bufferedReader = new BufferedReader(new FileReader(args[1]));
			}
			catch(FileNotFoundException e) {
				System.out.println("Il file non è stato trovato.\n");
				System.exit(1);
			}
		}

		try {
			while ((nread = bufferedReader.read()) >= 0) {
				read_char = (char) nread;
				for (int i = 0; i < characters_to_remove.length() && !charIsPresent; i++) {
					if (read_char == characters_to_remove.charAt(i)) {
						charIsPresent = true;
					}
				}

				if (!charIsPresent) {
					System.out.print(read_char);
				}

				charIsPresent = false;
			}

			System.out.print("\n");
			bufferedReader.close();
		}
		catch (IOException e) {
			System.out.println("Errore di input.\n");
			System.exit(2);
		}
	}
}

File: consumatore.c
#include <stdio.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#define MAX_STRING_LENGTH 256

int main(int argc, char* argv[]) {
	char read_char, charToRemove[MAX_STRING_LENGTH], carSeq;
	int nread, fd, trovato;
	
	strcpy(charToRemove, argv[1]);
	
	if (argc == 3) {
		fd = open(argv[2], O_RDONLY);
		if (fd < 0){
			perror("Impossibile aprire il file.");
			exit(1);
		}
	} else if (argc == 2) {
		fd = STDIN_FILENO;
	} else {
		perror("Numero di argomenti errato");
		exit(2);
	}
	
	while (nread = read(fd, &read_char, sizeof(char))) /* Fino ad EOF */ {
		if (nread < 0) {
			perror("Errore nella lettura del file");
			exit(3);
		}

		trovato = 0;
		
		for (int i = 0; !trovato && i < strlen(charToRemove); i++) {
			carSeq = charToRemove[i];

			if (read_char == carSeq) {
				trovato = 1;
			}	
		}

		if (!trovato) {
			putchar(read_char);
		}
	}

	close(fd);
}